samples/apps/autogen-studio/autogenstudio/workflowmanager.py

```diff
@@ -93,6 +94,75 @@ def __init__(
+                f = caller.register_for_llm(name=func_name, description=skill.description)(func)
+                executor.register_for_execution(name=func_name)(f)
+
+    def _connect_group_chat_skills(self, skills_module, group_configs, group_chat_manager):
+        agent_configs = group_configs.get("agents", [])
+        group = group_chat_manager.groupchat
+        admin_name = group.admin_name
+        admin_agent = group.agent_by_name(admin_name)
+        count_agents = len(agent_configs)
+        for i in range(count_agents):
+            agent_config = agent_configs[i]
+            agent_name = agent_config["config"].get("name")
+            agent = group.agent_by_name(agent_name)
+            # For some reason I don't yet understand, the group chat manager isn't given the correct name, so
+            # group.agent_by_name(agent_name) fails.
+            if agent == None:
+                # Ideally, every nested group within the group will have it's own unique name, but just in case...
+                groups = [agent for agent in group.agents
+                          if isinstance(agent, ExtendedGroupChatManager) and agent.manager_name == agent_name]
+                group_configs_list = [agent_config for agent_config in agent_configs if
+                                 agent_config.get("config", {}).get("name") == agent_name]
+                for j in range(len(groups)):
+                    self._connect_group_chat_skills(skills_module, group_configs_list[j], groups[j])
+            elif isinstance(agent, autogen.GroupChatManager):
+                self._connect_group_chat_skills(skills_module, agent_config, agent)
+            else:
+                if admin_agent:
+                    self._connect_agent_skills(skills_module, agent_config, agent, admin_agent)
+                else:
+                    # when setting up tool calls, you just need any other agent,
+                    # beside the current one.
+                    next_agent = i + 1
+                    if next_agent >= count_agents:
+                        next_agent = 0
+                    executor_name = agent_configs[next_agent].get("config").get("name")
+                    executor = group.agent_by_name(executor_name)
+                    self._connect_agent_skills(skills_module, agent_config, agent, executor)
+
+    def _connect_tools(self, sender_configs, receiver_configs):
+        import importlib.util
+        import sys
+
+        # import the skills as proper tools
+        # TODO: This may all be a bit too "wild west".
+        if self.work_dir not in sys.path:
+            sys.path.append(self.work_dir)  # TODO: Is it necessary?
+
+        spec = importlib.util.spec_from_file_location("skills", os.path.join(self.work_dir, "skills.py"))
+        skills_module = importlib.util.module_from_spec(spec)
+        spec.loader.exec_module(skills_module)
+        globals().update(vars(skills_module))  # TODO: Is it necessary?
```

Comment:
So these 4 lines are the biggest change here.  These lines import the "skills" file that gets saved into the currently running AutoGenStudio context.  I'm not sure if Python has a way to "sandbox" this code, but it's a pretty big deal because it the functions will get executed in the local context of the machine running AutoGenStudio.  Even if the user sets up Docker as their "code execution environment" that doesn't apply to tool calls.